crengine/src/lvtinydom.cpp

```
@@ -4937,14 +4945,17 @@ class LVImportStylesheetParser
                 lString32 css;
                 css << LVReadTextFile(cssStream);
                 int offset = _inProgress.add(cssFile);
-                ret = Parse(codeBase, css) || ret;
+                ret = Parse(codeBase, css, *styleSheet) || ret;
                 _inProgress.erase(offset, 1);
             }
         }
+
+        _document->getStyleSheet()->merge(*styleSheet);
+        cache.set(cssFile, styleSheet);
```

It's a bit hard to picture how all this lvtinydom.cpp parsing parts will play, given the various ways it can be used (html, epub, fb2), but it feels it is fine.
Except possibly at this point.
Depending on things (or even in a real stylesheet that uses `@import`), we may meet in `cssfile1.css`:
```
@import cssfile2.css;
p { ...}
```
When parsing cssfile1, we'll have nested calls with these 2 `LVImportStylesheetParser:Parse()` methods.
I think it is just fine if we end up caching the parsing result of `cssfile2.css`, and of `cssfile1.css` (not standalone, but including the content of `cssfile2.css`).
But it feels this `_document->getStyleSheet()->merge(*styleSheet);` may merge the same content multiple times into the real document _stylesheet ? ie. the main _stylesheet may get the selectors from css2file, then the selectors css1file (which includes the merged selectors of css2files), so finally having selectors in the order css2file|css2file|css1file ?
I'm not super sure this is happening, may be it's clearer to you (ie. you had thought about the possible mess I wrote and you're sure it's fine - or what I wrote rings something obvious on your side :) If the former, and it's all well thought of, feel free to write some comments explaining what's happening and why it is fine).

Dunno if this could be made clearer by just using/passing a temporary LVStyleSheet in the nested calls (keeping the temporary one you already added for caching) and do the merge into the Document real single _stylesheet in the `ldomDocument::parseStyleSheet()` after it has used LVImportStylesheetParser ?

---

Current implementation is indeed buggy. With your example, when `1.css` is first encountered, it goes to cache and merged to document stylesheet. `2.css` is also processed. The problem is that the cached `1.css` only contains its own rules, not those `@import`ed from `2.css`. The second time `1.css` is hit,  it's not parsed and `2.css` is silently ignored.

The fix is to introduce a `LVStyleSheet` parameter to `Parse(cssFile)`, so that nested stylesheets are parsed into it's (top) parent.

---

Thanks for confirming my hunch :)
Your updated code reads smooth (it feels right, simple, natural), so I guess it will work smooth :)